File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/flatland_space_stations.js
// https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true

function main() {
  const [nCities, _] = readNumbers();
  const lastCity = nCities - 1;
  let stations = readNumbers().sort((a, b) => a - b);
  console.log(maxDistanceFromSpaceStation(lastCity, stations));
}

function readNumbers() {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// n: length of stations
// T: O(n)
// S: O(1) extra space
function maxDistanceFromSpaceStation(lastCity, stations) {
  let maxDist = stations[0];

  for (let i = 1; i < stations.length; i++) {
    const gap = Math.trunc((stations[i] - stations[i - 1]) / 2);
    maxDist = Math.max(maxDist, gap);
  }

  return Math.max(maxDist, lastCity - stations.at(-1));
}

////////////////////////////////////////////////

const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines = [];
rl.on('line', (line) => inputLines.push(line));
rl.on('close', main);

function readLine() {
  return inputLines.shift();
}

File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/with morfo library/flatland_space_stations.java
// https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true
// Java 25

import morfo.io.Reader;

void main() {
    int nCities = Reader.readLnInts()[0];
    int lastCity = nCities - 1;
    int[] stations = Reader.readLnInts();
    Arrays.sort(stations);
    IO.println(maxDistanceFromSpaceStation(lastCity, stations));
}

// n: length of stations
// T: O(n)
// S: O(1) extra space
int maxDistanceFromSpaceStation(int lastCity, int[] stations) {
    int maxDist = stations[0];
    int n = stations.length;

    for (int i = 1; i < n; i++) {
        int gap = (stations[i] - stations[i - 1]) / 2;
        maxDist = Math.max(maxDist, gap);
    }

    return Math.max(maxDist, lastCity - stations[n - 1]);
}
